internal/jobs/jobobject.go

```diff
@@ -0,0 +1,363 @@
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.cpuRate != 0 || limits.cpuWeight != 0 || limits.cpuCount != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{}
+		if limits.cpuCount != 0 {
+			numProcessors := uint32(processorinfo.ProcessorCount())
+			if limits.cpuCount > numProcessors {
+				limits.cpuCount = numProcessors
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = calculateJobCPURate(limits.cpuCount, numProcessors)
+		} else if limits.cpuWeight != 0 {
+			if limits.cpuWeight < processorWeightMin || limits.cpuWeight > processorWeightMax {
+				return fmt.Errorf("processor weight value of `%d` is invalid", limits.cpuWeight)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+			cpuInfo.Value = calculateJobCPUWeight(limits.cpuWeight)
+		} else if limits.cpuRate != 0 {
+			if limits.cpuRate < processorLimitMin || limits.cpuRate > processorLimitMax {
+				return fmt.Errorf("processor rate of `%d` is invalid", limits.cpuRate)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = limits.cpuRate
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectCpuRateControlInformation, uintptr(unsafe.Pointer(&cpuInfo)), uint32(unsafe.Sizeof(cpuInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
+		}
+	}
+
+	if limits.maxBandwidth != 0 || limits.maxIops != 0 {
+		ioInfo := winapi.JOBOBJECT_IO_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE,
+		}
+		if limits.maxBandwidth != 0 {
+			ioInfo.MaxBandwidth = limits.maxBandwidth
+		}
+		if limits.maxIops != 0 {
+			ioInfo.MaxIops = limits.maxIops
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// CreateJobObject creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func CreateJobObject(name string) (*JobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
+	if err != nil {
+		return nil, err
+	}
+
+	iocpChan := make(chan IOCPNotif, 10)
+	jobMap[uint32(jobHandle)] = iocpChan
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	if ioCompletionPort == 0 {
+		ioInitOnce.Do(initIo)
+	}
+
```

Comment:
you could consider using a `recover` in this function to return an error more gracefully if initIo panics. 

Ref: https://blog.golang.org/defer-panic-and-recover

Reply:
I was thinking about this also. The iocp not being able to be made is unrecoverable but do you think there's any merit in just recovering and os.Exit() in the defer instead so its more "friendly"? We'd print the same error regardless, the user would just miss the gnarly stack trace. I decided it wasn't worth it but let me know

Reply:
I think what I'm concerned about is crashing containerd in this case. Someone could be running other container workloads and this would not be a good experience.

Reply:
How would the shim panicking crash containerd?

Reply:
Ah sorry, my mistake there 😅 But I think the point still stands. This could be undesired behavior, especially as we look to use job object containers in other ways.

Reply:
Came up with a nice middle ground offline with Kathryn. Can just make initIO inline in this function and if the iocp creation fails we can just assign a function local error to it. This way we can bubble up the error as we would normally